happi/audit.py

```diff
@@ -0,0 +1,424 @@
+        else:
+            # seems like the container has been validated
+            return self.report_code.SUCCESS
+
+    def validate_args(self, item):
+        """
+        Validates the args of an item
+        """
+        return [self.create_arg(item, arg) for arg in item.args]
+
+    def validate_kwargs(self, item):
+        """
+        Validates the kwargs of an item
+        """
+        return dict((key, self.create_arg(item, val))
+                    for key, val in item.kwargs.items())
+
+    def create_arg(self, item, arg):
+        """
+        Function borrowed from loader to create
+        correctly typed arguments from happi information
+        """
+        if not isinstance(arg, str):
+            return arg
+        return fill_template(arg, item, enforce_type=True)
+
+    def get_device_class(self, item):
+        """
+        Validates device_class field
+        """
+        device_class = item.get('device_class')
+        device = item.get('name')
+
+        if not device_class:
+            logger.warning('Detected a None vlaue for %s. '
+                           'The device_class cannot be None', device)
+            return self.report_code.MISSING
+        else:
+            try:
+                mod, cls = device_class.rsplit('.', 1)
+            except (Exception, ValueError) as e:
+                logger.warning('Wrong device name format: %s for %s, %s',
+                               device_class, device, e)
+                return self.report_code.INVALID
+            else:
+                packages = device_class.rsplit('.')
+                package = packages[0]
+                self._all_devices.add(package)
+                self._all_items.append(item)
+                return self._all_devices
+
+    def validate_import_class(self, item=None):
+        """
+        Validate the device_class of an item
+        """
+        device_class = item.get('device_class')
+        mod = device_class
+        if not device_class:
+            return self.report_code.MISSING
+        if '.' not in device_class:
+            logger.warning('Device class invalid %s for item %s',
+                           device_class, item)
+            return self.report_code.INVALID
+        else:
+            mod = device_class.rsplit('.')[0]
+        device = item.get('name')
+        if mod in self._all_devices:
+            # has been tested agains PyPi and was found there
+            try:
+                import_class(device_class)
+                return self.report_code.SUCCESS
+            except ImportError as e:
+                logger.warning('For device: %s, %s. Either %s is '
+                               'misspelled or %s is not part of the '
+                               'python environment', device, e,
+                               device_class, device_class)
+                return (self.report_code.INVALID, self.report_code.MISSING)
+        elif mod and mod not in self._all_devices:
```

Comment:
Is the `elif` necessary here? Isn't the only alternative it being invalid?

Reply:
This might not be necessary here if I don't have to deal with missing containers from pypi.... 
Also if I don't have to deal with searching packages in pypi I will probably be able to do a bit more cleanup here too